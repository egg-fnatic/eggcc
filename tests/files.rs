use std::path::PathBuf;

use eggcc::*;
use insta::assert_snapshot;
use libtest_mimic::Trial;

#[derive(Clone)]
struct Run {
    path: PathBuf,
}

impl Run {
    fn run(&self) {
        let program_name = self.path.file_stem().unwrap().to_str().unwrap();
        let program_read = std::fs::read_to_string(self.path.clone()).unwrap();
        let parsed = Optimizer::parse_bril(&program_read).unwrap();

        let mut optimizer_nothing = Optimizer::default().with_num_iters(0);
        let res_nothing = optimizer_nothing.optimize(&parsed).unwrap();

        assert_snapshot!(format!("{program_name}_no_opt"), format!("{}", res_nothing));

        let mut optimizer = Optimizer::default();
        let res = optimizer.optimize(&parsed).unwrap();

        // TODO test res and res_nothing to make sure
        // they evaluate the same as the original program
        // The cost of evaluating both should also go down
        // compared to original

        assert_snapshot!(format!("{program_name}"), format!("{}", res));
    }
}

fn generate_tests(glob: &str) -> Vec<Trial> {
    let mut trials = vec![];
    let mut mk_trial = |name: String, run: Run| {
        trials.push(Trial::test(name, move || {
            run.run();
            Ok(())
        }))
    };

    for entry in glob::glob(glob).unwrap() {
        let f = entry.unwrap();
        let name = f
            .file_stem()
            .unwrap()
            .to_string_lossy()
            .replace(['.', '-', ' '], "_");

        mk_trial(name.clone(), Run { path: f.clone() });
    }

    trials
}

fn main() {
    let args = libtest_mimic::Arguments::from_args();
    let tests = generate_tests("tests/**/*.bril");
    libtest_mimic::run(&args, tests).exit();
}
