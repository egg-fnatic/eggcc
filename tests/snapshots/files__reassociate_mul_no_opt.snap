---
source: tests/files.rs
assertion_line: 21
expression: "format!(\"{}\", res_nothing)"
---
@main {
  ret;
}
@mul_reassoc(x: int, y: int, z: int) {
  v0_: int = const 5;
  v1_: int = const 6;
  v2_: int = mul v1_ z;
  v3_: int = mul y v2_;
  v4_: int = mul v0_ v3_;
  v5_: int = mul x v4_;
  ret v5_;
}

